Next round of multireg preliminary changes

CarolEidt opened the pull request:
This is a zero-diff set of mostly refactoring changes in preparation for supporting multireg locals:
- Move `genRegCopy()` and `genStructReturn()` to codegencommon.cpp, making a new method for `genSIMDSplitReturn` which is target-specific.
- Factor out a new `genUnspillLocal` method from `genUnspillRegIfNeeded()`.
- Similarly factor out `genSpillLocal()`
- Rename `genMultiRegCallStoreToLocal()` and more generally support multireg local stores.
- Fix a bug in the order and shift amount for last-use bits on `GenTreeLclVar`
- Some additional cleanup and preparatory changes

CarolEidt committed:
Next round of multireg preliminary changes

This is a zero-diff set of mostly refactoring changes in preparation for supporting multireg locals:
- Move `genRegCopy()` and `genStructReturn()` to codegencommon.cpp, making a new method for `genSIMDSplitReturn` which is target-specific.
- Factor out a new `genUnspillLocal` method from `genUnspillRegIfNeeded()`.
- Similarly factor out `genSpillLocal()`
- Rename `genMultiRegCallStoreToLocal()` and more generally support multireg local stores.
- Fix a bug in the order and shift amount for last-use bits on `GenTreeLclVar`
- Some additional cleanup and preparatory changes

CarolEidt commented:
@dotnet/jit-contrib PTAL

CarolEidt commented on src/coreclr/src/jit/codegencommon.cpp:
```diff
@@ -11627,6 +11627,282 @@ void CodeGen::genReturn(GenTree* treeNode)
+    assert(!varTypeIsStruct(treeNode));
+    return false;
+#elif defined(TARGET_ARM64)
+    return varTypeIsStruct(treeNode) && (compiler->info.compRetNativeType == TYP_STRUCT);
```
On Arm64, SIMD types are returned in single registers, so they return false. Maybe this should be named `isMultiRegStructReturn()` and the logic may need some adjustment as we handle more struct types.

tannergooding replied:
Windows x64 should be in the same boat once https://github.com/dotnet/runtime/pull/32278 is completed.

CarolEidt commented on src/coreclr/src/jit/codegenarmarch.cpp:
```diff
@@ -1367,39 +1367,33 @@ void CodeGen::genPutArgSplit(GenTreePutArgSplit* treeNode)
-    GenTreeCall* call      = actualOp1->AsCall();
-    assert(call->HasMultiRegRetVal());
+    LclVarDsc* varDsc = compiler->lvaGetDesc(lclNum);
+    if (op1->OperIs(GT_CALL))
```
It can be another multi-reg local.

sandreenko replied:
I see only calls under `if (data->gtSkipReloadOrCopy()->IsMultiRegCall())`, so is it for your future changes?
The header should be updated `//    The child of the store is a multi-reg call or local?`

CarolEidt replied:
I'm going to update it simply to "The child of the store is a multi-reg node." since I hope that soon we will be supporting multi-reg returns from HW intrinsics.

sandreenko reviewed (approved):
LGTM, will check again when you delete TODO in headers.

Note: I have not checked functions that were moved to a different location, assuming there were no diffs there.

CarolEidt commented:
@sandreenko - I believe I've updated all the headers.

> Note: I have not checked functions that were moved to a different location, assuming there were no diffs there.

The only change to the functions I moved to codegencommon.cpp and codegenlinear.cpp was to extract the target-dependent code in `genSIMDSplitReturn()`

sandreenko commented on src/coreclr/src/jit/codegenlinear.cpp:
```diff
@@ -938,6 +938,84 @@ GenTree* sameRegAsDst(GenTree* tree, GenTree*& other /*out*/)
+{
+    LclVarDsc* varDsc = &compiler->lvaTable[varNum];
+    inst_set_SV_var(lclNode);
+    GetEmitter()->emitIns_R_S(ins_Load(type, compiler->isSIMDTypeLocalAligned(varNum)), emitTypeSize(type), regNum,
```
Nit, to avoid a strange new line with `	                              varNum, 0);`
```
    instruction ins = ins_Load(type, compiler->isSIMDTypeLocalAligned(varNum));
    GetEmitter()->emitIns_R_S(ins , emitTypeSize(type), regNum,  varNum, 0);
```

sandreenko commented on src/coreclr/src/jit/codegenlinear.cpp:
```diff
@@ -938,6 +938,84 @@ GenTree* sameRegAsDst(GenTree* tree, GenTree*& other /*out*/)
+void CodeGen::genUnspillLocal(
+    unsigned varNum, var_types type, GenTreeLclVar* lclNode, regNumber regNum, bool reSpill, bool isLastUse)
+{
+    LclVarDsc* varDsc = &compiler->lvaTable[varNum];
```
```suggestion
    LclVarDsc* varDsc = compiler->lvaGetDesc(varNum);
```

sandreenko commented on src/coreclr/src/jit/treelifeupdater.cpp:
```diff
@@ -126,15 +148,17 @@ void TreeLifeUpdater<ForCodeGen>::UpdateLifeVar(GenTree* tree)
+            for (unsigned i = 0; i < varDsc->lvFieldCnt; ++i)
             {
-                LclVarDsc* fldVarDsc = &(compiler->lvaTable[i]);
+                LclVarDsc* fldVarDsc = &(compiler->lvaTable[firstFieldVarNum + i]);
```
```suggestion
                LclVarDsc* fldVarDsc = compiler->lvaGetDesc(firstFieldVarNum + i);
```

sandreenko commented on src/coreclr/src/jit/codegencommon.cpp:
```diff
@@ -11627,6 +11627,282 @@ void CodeGen::genReturn(GenTree* treeNode)
+
+            if ((lcl->gtFlags & GTF_VAR_DEATH) == 0 && (treeNode->gtFlags & GTF_VAR_DEATH) == 0)
+            {
+                LclVarDsc* varDsc = &compiler->lvaTable[lcl->GetLclNum()];
```
```suggestion
                LclVarDsc* varDsc = compiler->lvaGetDesc(lcl);
```

sandreenko commented on src/coreclr/src/jit/codegenlinear.cpp:
```diff
@@ -985,99 +1064,26 @@ void CodeGen::genUnspillRegIfNeeded(GenTree* tree)
-            gcInfo.gcMarkRegPtrVal(dstReg, unspillTree->TypeGet());
+            bool reSpill   = ((unspillTree->gtFlags & GTF_SPILL) != 0);
+            bool isLastUse = lcl->IsLastUse(0);
+            genUnspillLocal(lcl->GetLclNum(), spillType, lcl->AsLclVar(), dstReg, reSpill, isLastUse);
```
Nit1: I think we probably do not need to pass both `lcl->GetLclNum()` and `lcl->AsLclVar()`.
Nit2: `lcl` is already a `LclVar`.

CarolEidt replied:
> I think we probably do not need to pass both lcl->GetLclNum() and lcl->AsLclVar()

We need both because when `lcl` is a multi-reg node we need to pass the node as well as the actual varnum of the specific field that's being spilled.

sandreenko replied:
Hm, I am missing something,
```
void CodeGen::genUnspillLocal(GenTreeLclVar* lclNode, var_types type, regNumber regNum, bool reSpill, bool isLastUse)
{
  unsigned varNum = lcl->GetLclNum();
```

won't it work?

CarolEidt replied:
A multi-reg local node will have the varNum of the parent. Each field has dedicated flags for last-use, spill and unspill, and they are in order of the fields' varNums (starting with `lvFieldLclStart`). For now (with no multi-reg nodes) these will be the same.

sandreenko replied:
Got it, thank you!

sandreenko commented on src/coreclr/src/jit/codegenlinear.cpp:
```diff
@@ -938,6 +938,84 @@ GenTree* sameRegAsDst(GenTree* tree, GenTree*& other /*out*/)
+//     varNum    - The variable number of the local to be spilled.
+//                 It may be a local field.
+//     type      - The type of the local.
+//     lclNode   - The node being spilled. Note that for a multi-reg local,
```
Nit: `being unspilled`

CarolEidt replied:
Also in comment on `varNum` - thanks!

sandreenko commented on src/coreclr/src/jit/codegenlinear.cpp:
```diff
@@ -938,6 +938,84 @@ GenTree* sameRegAsDst(GenTree* tree, GenTree*& other /*out*/)
 }
 
+//------------------------------------------------------------------------
+// genUnspillLocal: Reload a register candidate local into a register, if needed.
```
> `, if needed.`

Does this method decide if it is needed or not?

CarolEidt replied:
Ah, no it doesn't. The caller has to determine that. I'll change the comment.

sandreenko commented on src/coreclr/src/jit/gentree.h:
```diff
@@ -1733,6 +1733,9 @@ struct GenTree
     var_types GetRegTypeByIndex(int regIndex);
 
+    // Returns the GTF flag equivalent for the regIndex'th register of a multi-reg node.
+    unsigned int GetRegSpillFlagByIdx(int regIndex);
```
Nit: can it be const?

sandreenko reviewed (approved):
LGTM, one question about `genUnspillLocal` and a few nits.

CarolEidt commented:
The Windows_NT arm checked job failure is a time-out, and since this job passed prior to my last conflict-resolving commit, I'm going to go ahead and merge this.

jashook commented:
The test that timed out was:

```
Starting:    Loader.binding.XUnitWrapper
Collected crash dump C:\cores\crashdump_8496.dmp
    Loader\binding\tracing\BinderTracingTest.ResolutionFlow\BinderTracingTest.ResolutionFlow.cmd [FAIL]
      
      
      
      
      
      cmdLine:D:\h\w\B67E0984\w\AEAF09E1\e\Loader\binding\tracing\BinderTracingTest.ResolutionFlow\BinderTracingTest.ResolutionFlow.cmd Timed Out
      
      Return code:      -100
      Raw output file:      D:\h\w\B67E0984\w\AEAF09E1\e\Loader\binding\Reports\Loader.binding\tracing\BinderTracingTest.ResolutionFlow\BinderTracingTest.ResolutionFlow.output.txt
      Raw output:
      BEGIN EXECUTION
       "D:\h\w\B67E0984\p\corerun.exe" BinderTracingTest.ResolutionFlow.dll 
      Launching process for AppDomainAssemblyResolveEvent_Exception...
      Running AppDomainAssemblyResolveEvent_Exception...
      
      Launching process for LoadFromAssemblyPath_FoundInAppPath...
      Running LoadFromAssemblyPath_FoundInAppPath...
      
      Launching process for LoadFromAssemblyPath_FoundInLoadContext_DifferentMvid...
      Running LoadFromAssemblyPath_FoundInLoadContext_DifferentMvid...
      
      Launching process for LoadFromAssemblyPath_FoundInAppPath_IncompatibleVersion...
      Running LoadFromAssemblyPath_FoundInAppPath_IncompatibleVersion...
      
      Running AssemblyLoadContextResolving_ReturnNull...
      
      cmdLine:D:\h\w\B67E0984\w\AEAF09E1\e\Loader\binding\tracing\BinderTracingTest.ResolutionFlow\BinderTracingTest.ResolutionFlow.cmd Timed Out
      Test Harness Exitcode is : -100
      To run the test:
      > set CORE_ROOT=D:\h\w\B67E0984\p
      > D:\h\w\B67E0984\w\AEAF09E1\e\Loader\binding\tracing\BinderTracingTest.ResolutionFlow\BinderTracingTest.ResolutionFlow.cmd
      Expected: True
      Actual:   False
      Stack Trace:
        F:\workspace\_work\1\s\artifacts\tests\coreclr\Windows_NT.arm.Checked\TestWrappers\Loader.binding\Loader.binding.XUnitWrapper.cs(237,0): at Loader_binding._tracing_BinderTracingTest_ResolutionFlow_BinderTracingTest_ResolutionFlow_._tracing_BinderTracingTest_ResolutionFlow_BinderTracingTest_ResolutionFlow_cmd()
      Output:
        
        
        
        
        
        cmdLine:D:\h\w\B67E0984\w\AEAF09E1\e\Loader\binding\tracing\BinderTracingTest.ResolutionFlow\BinderTracingTest.ResolutionFlow.cmd Timed Out
        
        Return code:      -100
        Raw output file:      D:\h\w\B67E0984\w\AEAF09E1\e\Loader\binding\Reports\Loader.binding\tracing\BinderTracingTest.ResolutionFlow\BinderTracingTest.ResolutionFlow.output.txt
        Raw output:
        BEGIN EXECUTION
         "D:\h\w\B67E0984\p\corerun.exe" BinderTracingTest.ResolutionFlow.dll 
        Launching process for AppDomainAssemblyResolveEvent_Exception...
        Running AppDomainAssemblyResolveEvent_Exception...
        
        Launching process for LoadFromAssemblyPath_FoundInAppPath...
        Running LoadFromAssemblyPath_FoundInAppPath...
        
        Launching process for LoadFromAssemblyPath_FoundInLoadContext_DifferentMvid...
        Running LoadFromAssemblyPath_FoundInLoadContext_DifferentMvid...
        
        Launching process for LoadFromAssemblyPath_FoundInAppPath_IncompatibleVersion...
        Running LoadFromAssemblyPath_FoundInAppPath_IncompatibleVersion...
        
        Running AssemblyLoadContextResolving_ReturnNull...
        
        cmdLine:D:\h\w\B67E0984\w\AEAF09E1\e\Loader\binding\tracing\BinderTracingTest.ResolutionFlow\BinderTracingTest.ResolutionFlow.cmd Timed Out
        Test Harness Exitcode is : -100
        To run the test:
        > set CORE_ROOT=D:\h\w\B67E0984\p
        > D:\h\w\B67E0984\w\AEAF09E1\e\Loader\binding\tracing\BinderTracingTest.ResolutionFlow\BinderTracingTest.ResolutionFlow.cmd
```

jashook commented:
/cc @jaredpar
